### PR TITLE
filter cards using db queries instead of array methods

### PR DESCRIPTION
**Pull request by marvinRoellecke**
- add filter option to GET request in api/locations/index.js
- remove old filter array method in CardListGeneral.js
- using URLSearchParams.append / .delete to manipulate fetch url
- add states filterData, filterDataOther, cityFilter, sortData to keep filter settings when closing/reopening filterMenu
- adapt functions handleFilter and handleChangeSort to manipulate url and trigger fetch
- add functions handleFilterOther, handleCity

### CODE REVIEW

**Review by onebarloop — COMMENTED** [review]
Great work transforming your filter-logic to DB-querys! I left two small comments, nothing critical.

Keep it up, filterboy!

File: pages/api/locations/index.js [hunk 1]
```diff
@@ -5,7 +5,57 @@ export default async function handler(request, response) {
...
+
+    if (request.query.sort === "sortRating1") {
+      sort.rating = 1;
+    }
+
+    if (request.query.sort === "sortRating-1") {
+      sort.rating = -1;
+    }
+
+    const locations = await Location.find(filter).sort(sort);
```
Comment (julialoeschel) [inline pages/api/locations/index.js:58]: very elegant 👏

File: pages/api/locations/index.js [hunk 2]
```diff
@@ -5,7 +5,57 @@ export default async function handler(request, response) {
   await dbConnect();
 
   if (request.method === "GET") {
-    const locations = await Location.find();
+    const filter = {};
+
+    if (request.query.sport) {
+      filter["info.sport"] = request.query.sport;
+    }
+    if (request.query.rating) {
```
Comment (julialoeschel) [inline pages/api/locations/index.js:13]: can we filter this in the frontend ?

File: pages/index.js [hunk 3]
```diff
@@ -16,67 +16,123 @@ export default function Home({
...
-
-  //show / hide Filter Menu
+  const [filterDataOther, setFilterDataOther] = useState({
+    lighting: false,
+    wheelchair: false,
+    outdoor: false,
+    public: false,
+  });
+  const [cityFilter, setCityFilter] = useState();
+  const [sortData, setSortData] = useState();
```
Comment (julialoeschel) [inline pages/index.js:34]: always initiate an initial state here. 
```suggestion
  const [cityFilter, setCityFilter] = useState({});
  const [sortData, setSortData] = useState({});
```

File: pages/index.js [hunk 4]
```diff
@@ -16,67 +16,123 @@ export default function Home({
...
+  const [params, setParams] = useState(
+    new URL("http://localhost:3000/api/locations/?")
+  );
   const [isShowingFilterMenu, setIsShowingFilterMenu] = useState(false);
 
+  //show / hide Filter Menu
```
Comment (julialoeschel) [inline pages/index.js:40]: the name of the function is speaking by itself -> you don't need the comment here

File: pages/index.js [hunk 5]
```diff
@@ -16,67 +16,123 @@ export default function Home({
...
+      const response = await fetch("/api/locations/?" + searchURL);
+      const data = await response.json();
+      setLocations(data);
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  async function handleFilter(event, category) {
+    setFilterData((prevState) => {
```
[the remark this answers is not in the document]
Comment (marvinRoellecke) [reply pages/index.js:59]: yes that's right, I'm not using the prevState.
The callback functions creates a new const "newFilterData" that is used later on in line 71
`Object.entries(newFilterData).map((filterCategory) =>`
I removed the prevState as argument

**Comment by marvinRoellecke** [comment]
> Great work transforming your filter-logic to DB-querys! I left two small comments, nothing critical.
> 
> Keep it up, filterboy!

Thanks for the review :)

Filterboy 😂😂

**Review by julialoeschel — COMMENTED** [review]
Great Job Marvin, 
We viewed the PR together, you did a really great job  there. 
we have some comments for you. 

You could outsource some functions in index.js, as you proposed earlier. In this case you need to put the setter function as well. 

over all this is really advanced filter stuff, you can be proud! 
good job! 
🐜

**Pull request merged**